### PR TITLE
Pre-commit, darker, blacken-docs, flake8, pylint, bandit

### LINKED ISSUE #10932: Add pre-commit to code base
### Checklist

- [X] I added a descriptive title
- [X] I searched for other feature requests and couldn't find a duplicate (including also the ``type-feature`` tag)
- [X] I confirmed that it's not related to another project (see the above section)

### What is the idea?

We're testing code quality with flake8 right now, but don't have a good way to enforce code quality in local development environments. Instead of using a custom setup, let's use [pre-commit](https://pre-commit.com) for this and maybe even enable [pre-commit CI](https://pre-commit.ci) to get quicker results for PRs.

### Why is this needed

This is needed to streamline the development environment and enforce code quality checks for contributors before they are pushed to their forks.

### What should happen?

Add a pre-commit configuration and enable at least the flake8 hook.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by kenodegard**
Adds necessary [pre-commit](https://pre-commit.com/) config file.

Do we want to update the `README`/`CONTRIBUTING` in this PR to reflect these changes?

After internal brainstorming and discussions we decided on using the following CQ tools:
- [darker](https://github.com/akaihola/darker) (we want the formatting to be gradually applied)
- [blacken-docs](https://github.com/asottile/blacken-docs)
- [flake8](https://github.com/PyCQA/flake8) (this is what we have been using)

We also include two other useful tools which are run but we currently do not monitor the results, these results can be reviewed by running pre-commit in verbose mode:
- [pylint](https://github.com/PyCQA/pylint)
- [bandit](https://github.com/PyCQA/bandit)

Resolves #10932
CC: @rchord @jezdez

### CODE REVIEW

**Comment by dbast** [comment]
Nice! How about also changing the lint action and install `pre-commit` via `conda install conda-forge::pre-commit` and then running `pre-commit run -a` instead of flake8 at https://github.com/conda/conda/blob/master/.github/workflows/lint.yml#L31 ?

**Comment by dbast** [comment]
How about also activating some of the hooks of the default pre-commit repo?, e.g.
```
- repo: https://github.com/pre-commit/pre-commit-hooks
   rev: v4.0.1
   hooks:
   - id: mixed-line-ending
   - id: end-of-file-fixer
   - id: trailing-whitespace
```

If the pyproject.toml file only serves the purpose to provide parameters for black, while the underlying project is actually not pyproject.toml based (which is the case here), then the black parameters can also be set as pre-commit hook args e.g. 

```
-   repo: https://github.com/psf/black
    rev: 21.9b0
    hooks:
    -   id: black
        args: [-l, "99", --skip-magic-trailing-comma]
```

**Comment by kenodegard** [comment]
@dbast I hadn't modified our CI yet since we've been discussing using pre-commit CI.

I will look into those default hooks, I hadn't included them since trailing whitespace and end of line is fixed by black/darker so I was avoiding redundancy.

I disagree with removing the config file as that will make the config inaccessible to users who may choose to use these tools directly. I have no problem using a different config file but it doesn't look like black/darker accept anything besides `pyproject.toml`.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+[tool.black]
+line-length = 99
+target-version = ['py36', 'py37', 'py38']
+extend-exclude = '''
+# A regex preceded with ^/ will apply only to files and directories
+# in the root of the project.
+/(_vendor|devenv|tests)/
```
Comment (jezdez) [inline pyproject.toml:7]: I think the tests should be included in the move to use black.

```suggestion
/(_vendor|devenv)/
```

**Review by jezdez — APPROVED** [review]
This is a good start!

File: pyproject.toml (hunk 1 above)
Comment (kenodegard) [reply pyproject.toml:7]: Makes perfect sense, I don't remember why I included `tests` here...

Comment (kenodegard) [reply pyproject.toml]: Done

**Comment by kenodegard** [comment]
@jezdez do we want to enable pre-commit CI after merging this in or do we want to do as @dbast suggested and do `pre-commit run -a` in the Lint GH action?

**Comment by jezdez** [comment]
> @jezdez do we want to enable pre-commit CI after merging this in or do we want to do as @dbast suggested and do `pre-commit run -a` in the Lint GH action?

Let’s go with the service since we’re limited on number of runners already and we’ll soon add Python 3.10 which is added on top.

**Comment by jezdez** [comment]
That said, documenting how to use pre-commit locally is worth the effort for contributors.

**Comment by kenodegard** [comment]
Updated `CONTRIBUTING.md` with a static code analysis section and realized that I should add pre-commit to `tests/requirements.txt` that way it gets installed with `devenv`.

@jezdez I'll leave this for you to review one last time and merge as you see fit since you will also be able to enable pre-commit CI at the same time

**Review by jezdez — APPROVED** [review]
Love it!

**Pull request merged**